Feature/2 generate service annotations (#10)
cloudogu/k8s-dogu-operator: docs/operations/annotations_en.md

# Annotations for the k8s-dogu-operator

> This document contains the detailed description of all generated annotation of the k8s-dogu-operator.

Reviewer: ```suggestion
This document contains the detailed description of all generated annotations of the k8s-dogu-operator.
```

## Services

> This section contains all annotation that are appended to K8s services, if necessary.

Reviewer: ```suggestion
This section contains all annotations that are appended to K8s services, if necessary.
```

Author: The English document is automatically generated.

Reviewer: And this is the reason you should double-check it :D

### k8s-dogu-operator.cloudogu.com/ces-services

The `k8s-dogu-operator.cloudogu.com/ces-services` annotation contains information about one or more CES services.
Each CES service defines an external service of the system that is accessible through the web server. The annotation is automatically

> generated for each dogu that is marked as a webapp. It is also possible to customize the behavior of the services by specifying a

Reviewer: ```suggestion
generated for each dogu that is marked as a webapp. It is also possible to customize the behavior of the services by
```

Author: Done

> custom url by which the service is available. 

Reviewer: ```suggestion
a custom URL by which the service is available. 
```

Author: The English document is automatically generated.

**How to mark my dogu as webapp?**

> The requirement for you dogu is that the `Dockerfile` exposes at least one port. The dogu is marked as webapp via

Reviewer: ```suggestion
The requirement for your dogu is that the `Dockerfile` exposes at least one port. The dogu is marked as webapp via
```

Author: The English document is automatically generated.

> an environment variable. If the `Dockerfile` provides only one port, you have to set the environment variable

Reviewer: ```suggestion
an environment variable. If the `Dockerfile` provides only one port, you have to set the
```

Author: Done

`SERVICE_TAGS=webapp`. If the `Dockerfile` contains multiple ports it is required to mention the webapp's target port
in the environment variable. For example, we consider the exposed ports `8080,8081` and want to mark the port `8081` as
webapp. Then we need to set the environment variable `SERVICE_8081_TAGS=webapp` instead of `SERVICE_TAGS=webapp`.

**Example for a simple Webapp**

We consider the following setup:

Dockerfile
```
...
ENV SERVICE_TAGS=webapp
EXPOSE 8080
...
```

dogu.json
```yaml
...
"Name": "my-dogu-namespace/my-dogu-name"
...
```

The Dogu-Operator would create a service with the following `k8s-dogu-operator.cloudogu.com/ces-services` annotation:

```yaml
...
k8s-dogu-operator.cloudogu.com/ces-services: '[{"name":"my-dogu-name","port":8080,"location":"/my-dogu-name","pass":"/my-dogu-name"}]'
...
```

> Each `k8s-dogu-operator.cloudogu.com/ces-services` contains an arrays of `ces-service` JSON objects. Every `ces-service` 

Reviewer: ```suggestion
Each `k8s-dogu-operator.cloudogu.com/ces-services` entry contains an array of `ces-service` JSON objects. Every `ces-service` 
```

Author: The English document is automatically generated.

object contains the following information:
* name: The name of the ces service. Is used to identify the resulting ingress in the cluster.

> * port: The target port of the target service. In our case is the target service the generated dogu service.

Reviewer: ```suggestion
* port: The target port of the target service. In our case it is the target service of the generated dogu service.
```

Author: The English document is automatically generated.

> * location: The url where the ces-service is exposed to. Our ces-service would be available in the browser as 

Reviewer: ```suggestion
* location: The URL the ces-service is exposed to. Our CES-service would be available in the browser as 
```

Author: The English document is automatically generated.

  `http(s)://<fqdn>/my-dogu-name`.

> * pass: The url that should targeted in the target server.

Reviewer: ```suggestion
* pass: The URL that should be targeted in the target server.
```
How is this different from the `location`?

Author: The location is exposed. You type the location in your browser. The pass is used when the request is redirected to the target server. Sometimes it is necessary to change the context path of a request prior to sending it to the actual endpoint, e.g., the dogus are configured to listen to `localhost:<port>\context-path` instead of `localhost:<port>`

Author: Done

**Example for a Webapp with additional services**

Sometimes it is necessary to adapt the ces-service information or even add additional services for one dogu. The 

> following examples explains the necessary steps. We consider the following setup:

Reviewer: ```suggestion
following examples explain the necessary steps. We consider the following setup:
```

Author: The English document is automatically generated.

Dockerfile
```
...
ENV SERVICE_TAGS=webapp
ENV SERVICE_8080_NAME=superapp-ui
ENV SERVICE_8080_LOCATION=superapp
ENV SERVICE_ADDITIONAL_SERVICES='[{"name": "superapp-api", "port": 8080, "location": "api", "pass": "my-dogu-name/api/v2"}]'
EXPOSE 8080,8081
...
```

dogu.json
```yaml
...
"Name": "my-dogu-namespace/my-dogu-name"
...
```

The Dogu-Operator would create a service with the following `k8s-dogu-operator.cloudogu.com/ces-services` annotation:

```yaml
...
k8s-dogu-operator.cloudogu.com/ces-services: '[{"name":"superapp-ui","port":8080,"location":"/superapp","pass":"/my-dogu-name"},{"name":"superapp-api","port":8080,"location":"/api","pass":"/my-dogu-name/api/v2"}]'
...
```

The environment variables in the form `SERVICE_<PORT>_<PROPERTY>` allows overwrites of the default behaviour. As in our 
example the `SERVICE_8080_NAME=superapp-ui` makes the Dogu-Operator generate a ces service with the name `superapp-ui`
instead of `my-dogu-name` which is the name of the dogu. Accepted properties are `NAME`, `LOCATION`, and `PASS`.

> Besides, overwriting the default ces-service it is possible to add additional services. These are defined with the 

Reviewer: ```suggestion
Besides overwriting the default ces-service it is possible to add additional services. These are defined with the 
```

Author: The English document is automatically generated.
